worker/echo/main.go

```diff
@@ -40,39 +47,83 @@ func echo(
 		if sleepTime > 0 {
 			log.Infof("sleeping: %v", sleepTime)
 			time.Sleep(sleepTime)
 		}
-
-		// Set "response_to" according to the rcvId of the message we received
-		echoResponseTo := rcvId
-		// Create new echoId for the message we are going to send
-		echoId := uuid.New().String()
-
-		responseCode, responseMetadata, responseData, err := w.Transmit(
-			addr,
-			echoId,
-			echoResponseTo,
-			metadata,
-			data,
-		)
-		if err != nil {
-			return fmt.Errorf("cannot call Transmit: %w", err)
+		// Cancel message if it has been sent a cancel message
+		// during sleep time or during the loop
+		cancelChan, _ := sycMapCancelChan.Get(rcvId)
+		select {
+		case <-cancelChan:
+			// if the channel is close delete the channel from map
+			// it will not be longer used.
+			log.Tracef("canceled echo message id: %v", rcvId)
+			log.Tracef("Deleting channel from map")
+			sycMapCancelChan.Del(rcvId)
+			return nil
+		default:
+			if err := sendEchoMessage(w, addr, rcvId, responseTo, metadata, data, i); err != nil {
+				return err
+			}
 		}
 
-		// Log the responses received from the Dispatcher, if any.
-		log.Infof("responseCode = %v", responseCode)
-		log.Infof("responseMetadata = %#v", responseMetadata)
-		log.Infof("responseData = %v", responseData)
-		log.Infof("message %v of %v", i+1, loopIt)
+	}
 
-		if err := w.SetFeature("DispatchedAt", time.Now().Format(time.RFC3339)); err != nil {
-			return fmt.Errorf("cannot set feature: %w", err)
-		}
+	log.Tracef("Deleting channel from map")
+	sycMapCancelChan.Del(rcvId)
+	return nil
+}
+
+func sendEchoMessage(
+	w *worker.Worker,
+	addr string,
+	rcvId string,
+	responseTo string,
+	metadata map[string]string,
+	data []byte,
+	count int,
+) error {
+	// Set "response_to" according to the rcvId of the message we received
+	echoResponseTo := rcvId
+	// Create new echoId for the message we are going to send
+	echoId := uuid.New().String()
+
+	responseCode, responseMetadata, responseData, err := w.Transmit(
+		addr,
+		echoId,
+		echoResponseTo,
+		metadata,
+		data,
+	)
+	if err != nil {
+		return fmt.Errorf("cannot call Transmit: %w", err)
+	}
+
+	// Log the responses received from the Dispatcher, if any.
+	log.Infof("responseCode = %v", responseCode)
+	log.Infof("responseMetadata = %#v", responseMetadata)
+	log.Infof("responseData = %v", responseData)
+	log.Infof("message %v of %v", count+1, loopIt)
+
+	if err := w.SetFeature("DispatchedAt", time.Now().Format(time.RFC3339)); err != nil {
+		return fmt.Errorf("cannot set feature: %w", err)
+	}
+	return nil
+}
+
+// cancelEcho receives a cancel message id via  com.redhat.Yggdrasil1.Worker1.C.ancel method
+// closes the channel associated to that message to cancel its current run
+func cancelEcho(w *worker.Worker, addr string, id string, cancelID string) error {
+	log.Infof("cancelling message with id %v", cancelID)
+	if cancelChan, exists := sycMapCancelChan.Get(cancelID); exists {
+		close(cancelChan)
 	}
```

Comment:
When there is no message with given ID, then nothing happens. I would return error here. Something like this:

```go
else {
    return fmt.Errorf("message with given id: %v does not exist and cannot be canceled", cancelID)
)
```